### ISSUE TITLE
follow-hint-new-buffer(-focus) needs to login again in some sites

### ISSUE THREAD

**Issue by gdindi** [atlas-engineer/nyxt#1625]
**Describe the bug**
When logged in some sites (gitlab instances, exercism.io, for instance) using `follow-hint-new-buffer` or `follow-hint-new-buffer` opens the new buffer as if the user was not logged in.

**Precise recipe to reproduce the issue**

Go to https://gitlab.com and sign in. Use `follow-hint-new-buffer` or `follow-hint-new-buffer` and the new buffer will be open for a logged off session.

For website-specific issues:
Can you reproduce this issue with Epiphany / GNOME Web (https://wiki.gnome.org/Apps/Web)?

Open link in new tab works OK in Epiphany

**Information**
- OS name+version: Debian Bullseye
- Graphics card and driver: intel
- Desktop environment / Window manager name+version: EXWM
- How you installed Nyxt (Guix pack, package manager, build from source): built from source using debian packages only
- Information from command copy-system-information:

Nyxt version: 2.1.1-223-gd33257d4
Renderer version: GI-GTK
Operating system kernel: Linux 5.10.0-7-amd64
Lisp implementation: SBCL 2.1.1.debian (Dynamic space size: 1073741824)
Features: (WEBKIT2 WEBKIT2-2.32 WEBKIT2-PASTE-PLAINTEXT WEBKIT2-TRACKING WEBKIT2-MUTE
 WEBKIT2-EMOJI WEBKIT2-MEDIA WEBKIT2-SANDBOXING GTK-3-22 GTK-3-20 GTK-3-18
 GTK-3-16 GTK-3-14 GTK-3-12 GTK-3-10 GTK-3-8 GTK-3-6 GTK-3-4 GTK GDK-3-22
 GDK-3-20 GDK-3-18 GDK-3-16 GDK-3-14 GDK-3-12 GDK-3-10 GDK-3-8 GDK-3-6 GDK-3-4
 CAIRO-1-10 CAIRO-1-12 GDK-PIXBUF GLIB-2-30 GLIB-2-32 GLIB-2-34 GLIB-2-36
 GLIB-2-38 GLIB-2-40 GLIB-2-42 GLIB-2-44 GLIB-2-46 GLIB-2-48 GLIB-2-50
 GLIB-2-52 GLIB-2-54 GLIB-2-56 GLIB-2-58 GLIB NYXT-2 FSET-EXT-STRINGS
 CUSTOM-HASH-TABLE-NATIVE SWANK PLUMP-UTF-32 CL-PPCRE-UNICODE GLOBAL-VARS
 DECLARE-TYPES PARENSCRIPT SBCL+SAFE-STANDARD-READTABLE NAMED-READTABLES
 21BIT-CHARS CHUNGA GRAY-STREAMS FAST-IO-SV FAST-IO SBCL-USES-SB-ROTATE-BYTE
 ASDF-SYSTEM-CONNECTIONS CLOSER-MOP SPLIT-SEQUENCE CL-UNICODE FLEXI-STREAMS
 CL-PPCRE CL-JSON-DOUBLE-FLOAT-IS-SUBSUMED CL-JSON-SINGLE-FLOAT-IS-SUBSUMED
 CL-JSON-CLOS CL-JSON BORDEAUX-THREADS THREAD-SUPPORT FLAT-NAMESPACE X86-64
 UNIX CFFI FLAT-NAMESPACE SEQUENCE-EMPTYP QUICKLISP ASDF3.3 ASDF3.2 ASDF3.1
 ASDF3 ASDF2 ASDF OS-UNIX NON-BASE-CHARS-EXIST-P ASDF-UNICODE X86-64 GENCGC
 64-BIT ANSI-CL COMMON-LISP ELF IEEE-FLOATING-POINT LINUX LITTLE-ENDIAN
 PACKAGE-LOCAL-NICKNAMES SB-CORE-COMPRESSION SB-LDB SB-PACKAGE-LOCKS SB-THREAD
 SB-UNICODE SBCL UNIX)

ASDF version: 3.3.1
Critical dependencies: (/home/garjola/common-lisp/nyxt/_build/submodules/cl-cffi-gtk/gtk/cl-cffi-gtk.asd /home/garjola/common-lisp/nyxt/_build/quicklisp-client/dists/quicklisp/software/cl-gobject-introspection-20210124-git/cl-gobject-introspection.asd
 /home/garjola/common-lisp/nyxt/_build/submodules/cl-webkit/webkit2/cl-webkit2.asd)

Quicklisp dist version: 2021-06-30
Quicklisp client version: 2020-01-04
Local project directories: (/home/garjola/common-lisp/nyxt/_build/submodules/
 /home/garjola/common-lisp/nyxt/_build/quicklisp-client/local-projects/)
Critical dependencies(#<SYSTEM cl-cffi-gtk / cl-cffi-gtk-20201220-git / quicklisp 2021-06-30>
 #<SYSTEM cl-gobject-introspection / cl-gobject-introspection-20210124-git / quicklisp 2021-06-30>
 #<SYSTEM cl-webkit2 / cl-webkit-20210630-git / quicklisp 2021-06-30>)

Guix version: 1.2.0-19.8f9052d

**Comment by Ambrevar** [comment]
To clarify, does the issue happen if you simply open a new URL to the desired GitLab link?

**Comment by Ambrevar** [comment]
Sorry, I meant with `set-url-new-buffer`?

**Comment by gdindi** [comment]
Yes. It also happens if I manually set the url when I call the set-url-new-buffer command.

AFAIU, this happens when opening a new buffer, but not if one changes the url of the current buffer.

**Comment by Ambrevar** [comment]
I've had this problem with GitLab as well in the past, but recently the
issue has been that I would get logged out possibly more frequently than
expected.

Can you try deleting your cookies (default
~/.local/share/nyxt/cookies.txt) and start again?

**Comment by gdindi** [comment]
I just did that and the problem is still present.

**Comment by Ambrevar** [comment]
Can anyone else reproduce?

**Comment by jmercouris** [comment]
I can only reproduce it when I run multiple versions of WebKitGTK+. It seems that they interfere with each other and delete /overrwite cookies. Is there some tool that we can use that monitors the status of some files on disk and records the changes over time?

**Comment by Ambrevar** [comment]
John Mercouris ***@***.***> writes:

> when I run multiple versions of WebKitGTK+.

What do you mean exactly?

Multiple instances of Nyxt?

**Comment by jmercouris** [comment]
That's correct. I run multiple instances of Nyxt with different versions of WebKitGTK+ while developing.

**Comment by Ambrevar** [comment]
Do you mean "different WebKitGTK+ instances / processes"?
I'm guessing you are using WebKitGTK 2.32.*, right?

**Comment by Ambrevar** [comment]
I believe this should be fixed with https://github.com/atlas-engineer/nyxt/pull/1662. @gdindi Can you test Nyxt master and report if the issue is fixed or not for you

**Comment by gdindi** [comment]
On Thu 29-Jul-2021 at 09:36:58 +02, Pierre Neidhardt
***@***.***> wrote: 
> I believe this should be fixed with #1662. @gdindi Can you test Nyxt
> master and report if the issue is fixed or not for you? 

Hi,

I just compiled master and the problem persists (I tried with
gitlab.com) even after removing the cookies file.

--

**Comment by Ambrevar** [comment]
I can't reproduce anymore.

There was a bug on master by the time you tested which must have prevent
link hinting from following the URL altogether, so my guess is that if
you didn't encounter this issue, you are using an old version of Nyxt.

What does the `nyxt-version` command report?

If this is not the problem, can you try the latest commit and report an
exact recipe (with exact URLs), with all that you see?

**Comment by Ambrevar** [comment]
Also please try with `-I` or share your config files if it works with `-I`.

**Comment by gdindi** [comment]
Hi,
I just recompiled from master. 
```
(nyxt-version)
<INFO> [17:29:03] Version 2.1.1-418-g135155c1
```
The problem persists. To reproduce.
1. Go to https://gitlab.com/ambrevar/ambrevar.gitlab.io and log in
2. Follow hint in new buffer focus (C-u C-j with emacs bindings) and select for instance the «commits» button
3. I get the right page in a new buffer, but I am no longer loged in.

**Comment by gdindi** [comment]
What else can I do to help find the bug?

**Comment by jmercouris** [comment]
Please open the inspector in both buffers `open-inspector`, click the `storage` tab, and see if there are any cookies or removing of cookies that is unexpected. Try to figure out how to reliably reproduce the anomaly. I've attached a screenshot showing the cookies tab.

![Screenshot_20210730_210438](https://user-images.githubusercontent.com/1691662/127700284-25a81c5c-f46f-41c7-a85f-ce2bd2d919f9.png)

**Comment by gdindi** [comment]
The value of the cookie named _gitlab_session is different in both buffers. The buffer with the logged in session has cookies that are absent from the other one and also has cookies with expiration at the end of the session, while the buffer where I am not logged in has all cookies with expiration dates corresponding to a date and not to a session (which is normal, since I am not logged in).

I don't know anything about this stuff, but, for me, it is as if the cookies were not shared between the 2 buffers, and I supposed that this makes it impossible to be logged in in the second buffer.

I don't know if I am helping much, but tell me if there is anything else I can do.

**Comment by gdindi** [comment]
Just another info. With github (where follow-hint-new-buffer works) the cookie named __Host-user_session_same_site has the same value for both buffers. The same happens for the _device_id cookie. However, the _gh_sess cookie has different values.

For gitlab, there is a cookie with the same value in the 2 buffers and is called _sp_id.6b85. Otherwise, the buffer where I am logged in has a cookie named known_sign_in which does not exist in the other buffer.

**Comment by Ambrevar** [comment]
Could you try again on `master`, `2-series` or with the upcoming `2.2.3` see if this is fixed?

**Comment by gdindi** [comment]
Hi,

I just tested with master (Version 2.1.1-839-g76edb76f) and with version (2.2.3) and I have the same behaviour as before with both (I have only tested with 2 different instances of gitlab).
So the problem does not seem to be solved.

**Comment by jmercouris** [comment]
Thank you for the feedback. It seems we are still not doing something properly with WebKit's expectations for data storage. Perhaps some of the cookies are not getting copied over to the new buffer due to file locks?

**Comment by Ambrevar** [comment]
This should be fixed with #2050 now that it's on master.

**Comment by jmercouris** [comment]
Please reopen if this is not fixed.
